python/uv/repositories.bzl

```diff
@@ -0,0 +1,122 @@
+# Copyright 2024 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+EXPERIMENTAL: This is experimental and may be removed without notice
+
+Create repositories for uv toolchain dependencies
+"""
+
+load("@bazel_tools//tools/build_defs/repo:utils.bzl", "maybe")
+load("//python/uv/private:toolchain_types.bzl", "UV_TOOLCHAIN_TYPE")
+load("//python/uv/private:toolchains_repo.bzl", "uv_toolchains_repo")
+load("//python/uv/private:versions.bzl", "UV_PLATFORMS", "UV_TOOL_VERSIONS")
+
+UV_BUILD_TMPL = """\
+# Generated by repositories.bzl
+load("@rules_python//python/uv:toolchain.bzl", "uv_toolchain")
+
+uv_toolchain(
+    name = "uv_toolchain",
+    uv = "{binary}",
+    version = "{version}",
+)
+"""
+
+def _uv_repo_impl(repository_ctx):
+    platform = repository_ctx.attr.platform
+    uv_version = repository_ctx.attr.uv_version
+
+    is_windows = "windows" in platform
+
+    suffix = ".zip" if is_windows else ".tar.gz"
+    filename = "uv-{platform}{suffix}".format(
+        platform = platform,
+        suffix = suffix,
+    )
+    url = "https://github.com/astral-sh/uv/releases/download/{version}/{filename}".format(
+        version = uv_version,
+        filename = filename,
+    )
+    if filename.endswith(".tar.gz"):
+        strip_prefix = filename[:-len(".tar.gz")]
+    else:
+        strip_prefix = ""
+
+    repository_ctx.download_and_extract(
+        url = url,
+        sha256 = UV_TOOL_VERSIONS[repository_ctx.attr.uv_version][repository_ctx.attr.platform].sha256,
+        stripPrefix = strip_prefix,
```

Comment:
Nit, in order to ensure that this works correctly with internal indexes and the bazel downloader, it would be good to pass the `auth` headers, similar to how it is done in `whl_library.bzl`.